### PR TITLE
Bug 1907329: Add cluster profile support

### PR DESCRIPTION
**Pull request by guillaumerose**
Implementation of https://github.com/openshift/enhancements/pull/200

This adds a CLUSTER_PROFILE env variable that is used to determine which manifests to apply based on the `include.release.openshift.io/` annotation.

It is exactly what is written in the enhancement, not more.

This env. variable is not used by any components and, as it is empty, it defaults to the value of the default profile `self-managed-high-availability`.

### CODE REVIEW

**Comment by vrutkovs** [comment]
AWS failed to bootstrap

/retest

**Comment by guillaumerose** [comment]
@vrutkovs yes it is because I didn't solve the issue of "where we should store the profile". Currently, the CVO expects to find it in a ConfigMap and it is still created by hand. Do you think it's a good idea to read it from a ConfigMap ? There is a chicken and egg issue with this: if the ConfigMap is not present, the CVO cannot run but the ns openshift-config is created by the CVO.

**Comment by vrutkovs** [comment]
CVO should fall back to default profile (so that upgrade would succeed) if the configmap is not present

**Comment by guillaumerose** [comment]
OK, so I will change this PR to:
* load the ConfigMap as soon as the CVO starts.
* if the ConfigMap doesn't exist, use the default profile.

It will be the responsibility of the installer to have the ConfigMap in place before the CVO runs for the first time. Does it sounds good ?

**Comment by guillaumerose** [comment]
I experimented a bit more and found an interesting problem by making optional the ConfigMap.

During the boostrap, the CVO is created via a bootstrap-manifest and it runs before manifests are created (located in /manifests - even before the first one 0000_00_cluster-version-operator_00_namespace.yaml). It could not find the ConfigMap and therefore load the default profile.
Couple of minutes later, bootstrap ends, the ConfigMap is in place, the CVO reboot in the master and it's fine. But there was a period of time where the profile was not the good one. 

Any thoughts on this ? How can we make the CVO waiting for the ConfigMap but still having the choice to not have it ?
Isn't it a problem to have the CVO in the bootstrap node not running the right profile ?

**Comment by vrutkovs** [comment]
>But there was a period of time where the profile was not the good one.

I'm not sure I'm following - why installed profile is incorrect? I assumed that no profile specified == "default" profile. If its not set then all operators (unless explicitly excluded) are enabled

**Comment by guillaumerose** [comment]
I wanted to change the behavior to load once the profile at startup but it doesn't work. CVO is launched before manifests are created. At start the ClusterVersion object doesn't exist and locks the CVO. 

To work correctly with a profile, the user have to ensure the ConfigMap is created before the ClusterVersion. For snc, I did something like:

```
openshift-install create manifests

cp myns.yaml ${INSTALL_DIR}/openshift/0000_00_cluster-version-operator_00_myns.yaml
cp myconfigmap.yaml ${INSTALL_DIR}/openshift/0000_00_cluster-version-operator_00_myconfigmap.yaml

openshift-install create cluster
```

File: pkg/cvo/sync_worker.go
Comment (csrwng) [inline pkg/cvo/sync_worker.go:61]: add godoc

File: pkg/cvo/sync_worker.go
Comment (csrwng) [inline pkg/cvo/sync_worker.go:62]: You should return an error in this function. It's probably safer to fail in the case you can't read the profile than to assume "default"

File: pkg/cvo/updatepayload.go
Comment (csrwng) [inline pkg/cvo/updatepayload.go:334]: see comment for the interface

File: pkg/cvo/sync_worker.go
Comment (csrwng) [reply pkg/cvo/sync_worker.go:62]: Just to clarify, you can check the error for IsNotFound and return default... however other errors probably indicate an api server error and you should probably report.

File: pkg/cvo/sync_worker.go
Comment (guillaumerose) [reply pkg/cvo/sync_worker.go:62]: done :+1:

File: pkg/cvo/sync_worker.go
Comment (guillaumerose) [reply pkg/cvo/sync_worker.go]: yes you're right, it's better now.

File: pkg/cvo/updatepayload.go
Comment (guillaumerose) [reply pkg/cvo/updatepayload.go]: done :+1:

File: pkg/cvo/sync_worker.go
Comment (wking) [inline pkg/cvo/sync_worker.go:64]: I'm not a huge fan of `interface` that wrap a single function.  And I think the risk of needing to grow a new function in the future is small.  Can we make this:

```go
// RetrieveClusterProfile retrieves the cluster profile.
type RetrieveClusterProfile func(ctx context.Context) (string, error)
```

That would allow us to drop `fakeClusterProfileRetriever` and just inline `func (ctx context.Context) {return "default", nil}` where we needed a dummy.

File: pkg/cvo/sync_worker.go
Comment (wking) [reply pkg/cvo/sync_worker.go:62]: a84ec7a1e1 has a godoc; marking this thread resolved.

File: pkg/cvo/sync_worker.go
Comment (wking) [inline pkg/cvo/sync_worker.go:520]: I don't think we should `klog` here and then return the `err`.  The `syncOnce` caller is in charge of logging and otherwise announcing the error, as it does [here][1].  Can we drop this line?

[1]: https://github.com/openshift/cluster-version-operator/blob/a49fef5c66c6b0707c54fd93f84d2f51d3d28aca/pkg/cvo/sync_worker.go#L336

File: pkg/cvo/updatepayload.go
Comment (wking) [inline pkg/cvo/updatepayload.go:336]: Instead of hard-coding `openshift-config/cluster-profile` here, can we use `%s/%s`, `ConfigNamespace`, and `ClusterProfileConfigMap`?

File: pkg/internal/constants.go
Comment (wking) [inline pkg/internal/constants.go:7]: There are a few comments in the PR about if/how to use a ConfigMap to store this config (which was [out-of-scope][1] in the enhancement), starting [here][2].  Can we move them to this thread so they don't get interleaved with other PR-scoped comments?  I feel like managing this config may deserve its own enhancement :p.  Is this intended to be a install-time config?  What happens if admins twiddle the value post-install?  What happens if they set an invalid value (seems like the CVO blindly accepting an invalid profile would effectively cripple manifest application)?  Should we be reporting the configured profile through metrics/telemetry?  My initial knee-jerk is that this should be a ClusterVersion `status` property, to dodge some of these mutable-location wrinkles, but we're unlikely to be able to alter ClusterVersion without a formal enhancement.  Thoughts?

[1]: https://github.com/openshift/enhancements/pull/200#discussion_r394654607
[2]: https://github.com/openshift/cluster-version-operator/pull/404#issuecomment-662516200

File: pkg/payload/payload_test.go
Comment (wking) [inline pkg/payload/payload_test.go:149]: this is a weird test-case.  I'd expect an empty-string profile to cause `shouldExclude` to return an error (which would require us to update its call signature) or panic or some-such.  If we happen to get into that situation, loading a bunch of competing manifests from different profiles does not seem like something that would work well.

File: pkg/payload/payload_test.go
Comment (wking) [inline pkg/payload/payload_test.go:194]: nit: useless empty space.  Can we drop this line?

File: pkg/payload/payload_test.go
Comment (guillaumerose) [reply pkg/payload/payload_test.go:149]: Yes it's a problem.
At the initialization of the CVO, it loads the payload to get the verifier ConfigMap (https://github.com/openshift/cluster-version-operator/pull/404/files#diff-85c31beb4341b4c52f892c0581bbb5d6R226) but don't have access to the profile. At this point of the install, the CVO container is mostly alone and no resources are present. Either we wait for the ConfigMap (but it's optional, so how can we do that ?) or we load all or default manifests and take the first verifier.
In the sync loop, it works correctly because the ClusterVersion object acts as a lock and the ConfigMap cluster-profile is created before.

File: pkg/internal/constants.go
Comment (guillaumerose) [reply pkg/internal/constants.go:7]: For CRC, it is definitely an install-time config and we don't want the admin to modify it.
In this PR, the ClusterVersion object is key and the ConfigMap with the profile must be created before (as the CV is a lock for the sync loop). The admin have to insert the correct manifest between the `openshift-install create manifests` and `openshift-install create cluster` phases.  
I must admit it would cleaner if we can have a profile property in the CV like CVO overrides.
Is this first version acceptable still? WDYT?

File: pkg/cvo/sync_worker.go
Comment (guillaumerose) [reply pkg/cvo/sync_worker.go]: done :+1:

File: pkg/cvo/sync_worker.go
Comment (guillaumerose) [reply pkg/cvo/sync_worker.go]: done :+1:

File: pkg/cvo/updatepayload.go
Comment (guillaumerose) [reply pkg/cvo/updatepayload.go]: done :+1:

File: pkg/payload/payload_test.go
Comment (guillaumerose) [reply pkg/payload/payload_test.go]: done :+1:

File: pkg/payload/payload_test.go
Comment (wking) [reply pkg/payload/payload_test.go:149]: >  Either we wait for the ConfigMap (but it's optional, so how can we do that ?) or we load all or default manifests and take the first verifier.

"verifier"?  The enhancement line I comment on [here][1] suggests the no-configured-profile behavior should be "exclude manifests with any profile annotations".  That also mitigates the config-loading race, by pushing no profile-specific manifests until a profile config has been found.

[1]: https://github.com/openshift/enhancements/pull/200#discussion_r457408715

File: pkg/payload/payload.go
Comment (wking) [inline pkg/payload/payload.go:204]: Per [this][1], we should move the empty-profile check down and have:

```go
if !shouldCheckProfile{
  return false // include manifests which declare no profiles
}

if profile == "" {
  return true // exclude manifests which declare profiles when no profile is configured
}

profileAnnotation := fmt.Sprintf("include.release.openshift.io/%s", profile)
...
```


[1]: https://github.com/openshift/cluster-version-operator/pull/404#discussion_r461551007

File: pkg/cvo/updatepayload.go
Comment (wking) [inline pkg/cvo/updatepayload.go:322]: This needs to check for [the `CLUSTER_PROFILE` environment variable][1] before looking at ConfigMaps or other locations.

[1]: https://github.com/openshift/enhancements/pull/200/files#diff-b0d3a9e089e5cf5a125a408f84f934abR67

File: pkg/cvo/updatepayload.go
Comment (wking) [inline pkg/cvo/updatepayload.go:326]: Per [this][1], I think we shoulf `return "", nil` here.

[1]: https://github.com/openshift/cluster-version-operator/pull/404#discussion_r461555262

File: pkg/cvo/updatepayload.go
Comment (wking) [inline pkg/cvo/updatepayload.go:325]: No need to format `err` in here.  We already know it's not-found and we can express that in English like `ConfigMap %s/%s not found, excluding all profiles`.

File: pkg/payload/payload_test.go
Comment (guillaumerose) [reply pkg/payload/payload_test.go:149]: The verifier is created at this point using an update: https://github.com/openshift/cluster-version-operator/blob/master/pkg/cvo/cvo.go#L245.
This is done once at the initialization of the CVO, so we can't wait for the profile to appear I suppose.

Comment (wking) [reply pkg/payload/payload_test.go:149]: Why would you need profile-specific verification?

Comment (wking) [reply pkg/payload/payload_test.go:149]: > This is done once at the initialization of the CVO...

We probably need to cache the profile and reload from disk if it's changed since the last sync cycle.

Comment (guillaumerose) [reply pkg/payload/payload_test.go:149]: > Why would you need profile-specific verification? 

You're right. I don't want it. 
At initialization, we can load default profile manifests and take the verifier. It removes this empty string profile and simplify `shouldExclude`. I will give it a try a do that.

> We probably need to cache the profile and reload from disk if it's changed since the last sync cycle.

Cache the profile is I think a good thing. Can the informer notify when a change appear?

Comment (wking) [reply pkg/payload/payload_test.go]: > Can the informer notify when a change appear?

The ConfigMap informer could, but it would make a generic interface/getter more awkward to expose that.  I'd rather just poll once at the start of each sync cycle.

File: pkg/cvo/updatepayload.go
Comment (guillaumerose) [reply pkg/cvo/updatepayload.go]: done

File: pkg/cvo/updatepayload.go
Comment (guillaumerose) [reply pkg/cvo/updatepayload.go]: done! Now there is no more mention of the default profile, it's simpler.

File: pkg/cvo/updatepayload.go
Comment (guillaumerose) [reply pkg/cvo/updatepayload.go:325]: added!

File: pkg/payload/payload.go
Comment (guillaumerose) [reply pkg/payload/payload.go]: done

Comment (guillaumerose) [reply pkg/payload/payload.go]: I just realized that with this change we can only add more manifests but not replace or deactivate default manifests. Let's say I want a different image for the ingress operator. One manifest will have the include annotation, the other will not have it. If I run the CVO with the profile, the 2 manifests will be applied.
How can we avoid this ?

File: pkg/cvo/updatepayload.go
Comment (LalatenduMohanty) [reply pkg/cvo/updatepayload.go:325]: @guillaumerose Where is the env variable?

Comment (guillaumerose) [reply pkg/cvo/updatepayload.go:325]: Here https://github.com/openshift/cluster-version-operator/pull/404/files#diff-4f43e557624e0fc47bc4e31f4f236ab5R99

Comment (LalatenduMohanty) [reply pkg/cvo/updatepayload.go:325]: I do not see the env viariable check before looking at ConfigMaps

File: pkg/internal/constants.go
Comment (LalatenduMohanty) [reply pkg/internal/constants.go:7]: @guillaumerose Why do you think CRC need it as install-time config? To me it should work for CRC even if you provide the ConfigMap post the first installation.

File: pkg/cvo/updatepayload.go
Comment (guillaumerose) [reply pkg/cvo/updatepayload.go:325]: It's weird it doesn't appear here. If I browse the `Files changed` tab I can see it: https://github.com/openshift/cluster-version-operator/pull/404/files#diff-d540a41404f678a1c438f4c1e5b92a87R322-R325

File: pkg/internal/constants.go
Comment (guillaumerose) [reply pkg/internal/constants.go:7]: It should work post-install but it will take extra time (10-20min?) to install unnecessary components. It can also have weird side effects. It would be great to guaranty that a profile is not dependent of the default profile.

Comment (LalatenduMohanty) [reply pkg/internal/constants.go:7]: Any change to API also impact to multiple components (Installer, CVO) should be an enhancement https://github.com/openshift/enhancements#is-my-thing-an-enhancement

Comment (guillaumerose) [reply pkg/internal/constants.go:7]: To move this PR forward, I propose to:
* remove the use of the ConfigMap,
* only use the env. variable (as stated in the enhancement).
* open a PR in the installer to add the env variable.

WDYT?

File: pkg/payload/payload.go
Comment (wking) [reply pkg/payload/payload.go]: > Let's say I want a different image for the ingress operator...

Open a new PR to adjust the enhancement and say:

> Manifests with no explicit inclusion annotations implicitly belong to the `default` profile.  If no profile is configured, the cluster-version operator should use the `default` profile.

File: pkg/internal/constants.go
Comment (wking) [reply pkg/internal/constants.go:7]: You'd need to adjust the `render` logic here and also pur default CVO Deployment templating to propagate the env var between updates.  But sounds plausible to me.

File: pkg/payload/payload.go
Comment (guillaumerose) [reply pkg/payload/payload.go]: PR opened https://github.com/openshift/enhancements/pull/414

File: pkg/internal/constants.go
Comment (guillaumerose) [reply pkg/internal/constants.go]: I updated the PR with only the env variable and the flag. It's now clearer although the flag `--cluster-profile` is a little bit invasive. Let me know what you think.

File: cmd/render.go
Comment (wking) [inline cmd/render.go:45]: It should default to `default`, not be required.  And I see you set `default` above when declaring the `StringVar`.  So the error message should probably be `"invalid empty --cluster-profile value"` or some such.

File: pkg/cvo/cvo_scenarios_test.go
Comment (wking) [inline pkg/cvo/cvo_scenarios_test.go:97]: This should get set in `&Operator` above and then passed through here.

File: cmd/start.go
Comment (wking) [inline cmd/start.go:35]: `start.NewOptions()` should be defaulting this to `default`.  But I'm not seeing that code anywhere.  Am I just missing it?

File: pkg/payload/payload.go
Comment (wking) [inline pkg/payload/payload.go:206]: simpler implementation might be:

```go
prefix := "include.release.openshift.io/"
exists := struct{}{}
profiles := map[string]struct{}{}
for key, value := range annotations {
  if strings.HasPrefix(key, prefix) && value == "true" {
    profiles[key[len(prefix):]] = exists
  }
}

if len(profiles) == 0 {
  profiles["default"] = exists
}

_, belongsInProfile := profiles[profile]
return !belongsInProfile
```

File: cmd/start.go
Comment (wking) [reply cmd/start.go:35]: Never mind, found it with the `defaultClusterProfile` stuff.

File: pkg/payload/payload_test.go
Comment (wking) [inline pkg/payload/payload_test.go:154]: add a test case where `profile: "default"` shows us excluding `annotations: map[string]interface{}{"include.release.openshift.io/edge": "true"}`, i.e. that the manifest only belongs implicitly to `default` if it declares no explicit profile inclusion.

File: pkg/payload/payload_test.go
Comment (wking) [inline pkg/payload/payload_test.go:157]: this should exclude the manifest, because the manifest is implicitly a member of the `default` profile, and not a member of the `edge` profile.

Comment (guillaumerose) [reply pkg/payload/payload_test.go:157]: This case ensures we don't have to duplicate all manifests for all profiles. We can probably change the enhancemen to say 
`Manifests with no explicit inclusion annotations implicitly belong to all profiles. `.

File: pkg/cvo/cvo_scenarios_test.go
Comment (guillaumerose) [reply pkg/cvo/cvo_scenarios_test.go]: done :+1:

File: pkg/payload/payload_test.go
Comment (guillaumerose) [reply pkg/payload/payload_test.go:159]: done :+1:

File: cmd/render.go
Comment (guillaumerose) [reply cmd/render.go]: done :+1:

File: bootstrap/bootstrap-pod.yaml
Comment (guillaumerose) [inline bootstrap/bootstrap-pod.yaml:20]: We might want to add a condition like this. It will prevent to add the default value in the manifest and spread this flag too much.
``` 
{{if ne .ClusterProfile "default" }}
      - "--cluster-profile={{.ClusterProfile}}"
{{end}}
```
WDYT @wking ?

Comment (wking) [reply bootstrap/bootstrap-pod.yaml:20]: I'm fine as this stands.  One less place to hard-code `default`

File: pkg/payload/payload_test.go
Comment (wking) [reply pkg/payload/payload_test.go:162]: > This case ensures we don't have to duplicate all manifests for all profiles.

David [wants][1] that legwork to be a new-profile entry barrier.

[1]: https://github.com/openshift/enhancements/pull/200#discussion_r375837903

**Comment by bbrowning** [comment]
This looks like a real failure in that e2e-upgrade test:
```
fail [github.com/openshift/origin@/test/e2e/upgrade/upgrade.go:139]: during upgrade to registry.svc.ci.openshift.org/ci-op-jstn9hsr/release@sha256:64f525746fcc9663be6c4b70bb4e4d98d9f8d57fa5db6c90b0de8072869ce761
Unexpected error:
    <*errors.errorString | 0xc00208be10>: {
        s: "Cluster did not complete upgrade: timed out waiting for the condition: Error loading manifests from /etc/cvo/updatepayloads/QGnAQsjGa670WqA9XFnWRg: error running preprocess on 0000_00_cluster-version-operator_03_deployment.yaml: failed to execute template: template: manifest:32:33: executing \"manifest\" at <.ClusterProfile>: can't evaluate field ClusterProfile in type payload.manifestRenderConfig",
    }
```

It's specifically hitting issues around the `ClusterProfile` field.

**Comment by bbrowning** [comment]
My reading of the code and upgrade error here is that the previous version of the CVO needs to have the `ClusterProfile` field added to the manifestRenderConfig struct BEFORE that field is ever used in an actual manifest. During an upgrade, it looks like the existing CVO reads the upgrade payload to validate it and it's failing because the existing CVO doesn't know how to handle this new `ClusterProfile` field.

I think that means there are a couple of options:
1. Release this change over two releases, where the first release adds the new `ClusterProfile` field but does not reference it in any manifest. Then, in a later release, it can be referenced from upgrade manifests.
2. Release this change in one release, but explicitly do NOT reference this field in any manifest. That means the logic of what profile to default to when none is specified may have to get spread around a bit more.

I'm not super familiar with the mechanics of how CVO operates here, and the above is just based on a quick reading of the failing test and code. Perhaps someone else in this thread can verify the above is correct.

**Comment by guillaumerose** [comment]
Thanks @bbrowning, I didn't see the failure.

I guess we have 3 options:
* Continue with the env. variable/new parameter. First release, we define the new variable in the templating, second release we use it. It still requires some changes in the installer.
* Use a ConfigMap. No changes are needed in the installer but an enhancement PR should define it.
* Add a new field in ClusterVersion. We would need also an enhancement PR.

WDYT @abhinavdahiya @wking ?

**Comment by derekwaynecarr** [comment]
Is there an enhancement that references how CRC will use cluster profiles?  The IBM Cloud managed service case is clear since its already a production service and the alternative/excluded manifests in that setting is clearly understood.

**Comment by LalatenduMohanty** [comment]
@guillaumerose I guess we are still planning for this PR in 4.7, right?

**Comment by guillaumerose** [comment]
@LalatenduMohanty I guess, yes. The major issue is that I/we don't know how we should implement it. We need to make a choice between the 3 solutions here https://github.com/openshift/cluster-version-operator/pull/404#issuecomment-668557003 of if someone has an other idea.

**Comment by rkukura** [comment]
> Is there an enhancement that references how CRC will use cluster profiles? The IBM Cloud managed service case is clear since its already a production service and the alternative/excluded manifests in that setting is clearly understood.
> 
> see: [openshift/enhancements#418](https://github.com/openshift/enhancements/pull/418)

The enhancement PR for a single-node-developer cluster profile for CRC is: [openshift/enhancements#482](https://github.com/openshift/enhancements/pull/482)

**Comment by guillaumerose** [comment]
I reduced the scope of this PR to the minimal possible set of changes. It is linked to this enhancement PR: https://github.com/openshift/enhancements/pull/414

Before merging it, we would need all subcomponents to have the proper include annotation (see bug 1871890). 

This changes only addresses what is required for IBM Cloud. For single node profile, we still need to bring this option to the installer.

File: pkg/cvo/cvo.go
Comment (cfergeau) [inline pkg/cvo/cvo.go:228]: The  default name is `self-managed-high-availability`

Comment (guillaumerose) [reply pkg/cvo/cvo.go]: fixed, thanks

**Comment by guillaumerose** [comment]
Rebased.

@LalatenduMohanty PTAL

**Comment by eparis** [comment]
The unit test failure looks like a flake. But the lack of manifest getting applied, thus breaking cluster reader for insights operator does seem to be specific and repeatable to this PR.

**Comment by wking** [comment]
/hold

Blocked by operator-framework/operator-lifecycle-manager#1887 to label the InstallPlans CRD manifest (among other things) to avoid [`cannot list resource "installplans" in API group "operators.coreos.com" in the namespace "openshift"`][1].

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/404/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic/1338647777357336576#1:build-log.txt%3A73

**Review by LalatenduMohanty — APPROVED** [review]
/lgtm

**Comment by wking** [comment]
/hold cancel

I'd [held this][1] for operator-framework/operator-lifecycle-manager#1887, which has since landed.

[1]: https://github.com/openshift/cluster-version-operator/pull/404#issuecomment-746662144

**Review by LalatenduMohanty — APPROVED** [review]
/lgtm

**Pull request merged**
